### PR TITLE
feat: Config updater for the base provider

### PR DESCRIPTION
**Pull request by davidterpay**
closes BLO-901

This PR introduces a `ConfigUpdater` interface that allows the base provider to be updated asynchronously. Anytime a configuration update is received, this will trigger the provider to restart with the new configurations.

### CODE REVIEW

File: providers/base/config.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"context"
+
+	"go.uber.org/zap"
+)
+
+// listenConfigUpdater listens for updates from the config updater and updates the
+// provider's internal configurations. This will trigger the provider to restart
+// and is blocking until the context is cancelled.
+func (p *Provider[K, V]) listenConfigUpdater(ctx context.Context) {
```
Comment (aljo242) [inline providers/base/config.go]:
```suggestion
func (p *Provider[K, V]) listenOnConfigUpdater(ctx context.Context) {
```

File: providers/base/provider.go [hunk 2]
```diff
@@ -90,15 +97,66 @@ func (p *Provider[K, V]) Start(ctx context.Context) error {
...
+		select {
+		case <-p.restartCh:
+			// If any of the provider's configurations have changed, the provider will
+			// be signalled to restart.
+			p.logger.Info("restarting provider")
+			cancel()
+
+			// Wait for the fetch loop to stop.
+			err := <-errCh
+			p.logger.Debug("provider fetch loop stopped", zap.Error(err))
```
Comment (aljo242) [inline providers/base/provider.go:132]:
```suggestion
Restart:
	for {
		// Create a new context for the fetch loop. This allows us to cancel the fetch loop
		// when the provider needs to be restarted.
		fetchCtx, cancel := context.WithCancel(ctx)
		defer cancel()

		// Start the fetch loop.
		errCh := make(chan error)
		go func() {
			errCh <- p.fetch(fetchCtx)
		}()

		select {
		case <-p.restartCh:
			// If any of the provider's configurations have changed, the provider will
			// be signalled to restart.
			p.logger.Info("restarting provider")
			cancel()

			// Wait for the fetch loop to stop.
			err := <-errCh
			p.logger.Debug("provider fetch loop stopped", zap.Error(err))
			break Restart
```

I think this code is a bit hard to reason about, so adding this label (while redundant) makes things a bit clearer.

File: providers/base/fetch.go [hunk 3]
```diff
@@ -80,7 +76,8 @@ func (p *Provider[K, V]) startAPI(ctx context.Context, responseCh chan<- provide
 // attemptAPIDataUpdate tries to update data by fetching and parsing API data.
 // It logs any errors encountered during the process.
 func (p *Provider[K, V]) attemptAPIDataUpdate(ctx context.Context, responseCh chan<- providertypes.GetResponse[K, V]) {
-	if len(p.ids) == 0 {
+	ids := p.GetIDs()
```
Comment (g3ndo9kar9) [inline providers/base/fetch.go:79]: Unclear what p.GetIDs() is?

File: providers/base/utils.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	providertypes "github.com/skip-mev/slinky/providers/types"
+)
+
+type (
+	// The ConfigUpdater is an interface that be utilized to fetch the configuration
+	// for a given provider. This allows the provider to be updated asynchronously.
+	ConfigUpdater[K providertypes.ResponseKey] interface {
+		// GetIDs is the channel that is used to update the set of IDs that the provider
+		// will fetch data for. This blocks until there is a viable update.
```
Comment (g3ndo9kar9) [inline providers/base/utils.go:12]: Any reason to choose a channel v. a RWMutex here? Don't you want all Providers to read the updated ids concurrently? Do we expect a ConfigUpdater per Provider?

File: providers/base/provider.go [hunk 5]
```diff
@@ -90,15 +97,66 @@ func (p *Provider[K, V]) Start(ctx context.Context) error {
...
+		case err := <-errCh:
+			return err
+		case <-ctx.Done():
+			return <-errCh
+		}
```
Comment (g3ndo9kar9) [inline providers/base/provider.go:143]: Why do we need to kill the providers if there is a config update? Don't we j need to update the set of IDs it's fetching for?

Comment (aljo242) [reply providers/base/provider.go:143]: It works cleanly for that with REST providers, but for websockets, things get more complicated.  We would need to do unsubscribe/subscribe logic based on the diff we get in providers. 

This just simplifies things, creating a new connection so we have a fresh set of subscriptions.

File: providers/base/utils.go (hunk 4 above)
Comment (aljo242) [reply providers/base/utils.go:12]: Right now, this is being done per provider

File: providers/base/fetch.go (hunk 3 above)
Comment (davidterpay) [reply providers/base/fetch.go:79]: IDs are effectively currency pairs, the base provider just calls them IDs to have some generality.

File: providers/base/utils.go (hunk 4 above)
Comment (davidterpay) [reply providers/base/utils.go:12]: yea the interface that the config client, which is the client that is responsible for updating all providers, will use is the ConfigUpdater - which is specific per provider.

File: providers/base/provider.go [hunk 6]
```diff
@@ -90,15 +97,78 @@ func (p *Provider[K, V]) Start(ctx context.Context) error {
...
+		select {
+		case <-p.restartCh:
+			// If any of the provider's configurations have changed, the provider will
+			// be signalled to restart.
+			p.logger.Info("restarting provider")
+			cancel()
+
+			// Wait for the fetch loop to stop.
+			err := <-errCh
+			p.logger.Debug("provider fetch loop stopped", zap.Error(err))
```
Comment (aljo242) [inline providers/base/provider.go:132]:
```suggestion
			p.logger.Debug("provider fetch loop stopped", zap.Error(err))
			continue MainLoop
```

**Review by aljo242 — APPROVED** [review]

**Pull request merged**